static/src/javascripts/projects/commercial/modules/dfp/performance-logging.js

```diff
@@ -0,0 +1,199 @@
+// @flow
+
+/* eslint no-param-reassign: "off" */
+
+import raven from 'lib/raven';
+import config from 'lib/config';
+import { getCurrentTime } from 'lib/user-timing';
+import beacon from 'common/modules/analytics/beacon';
+import ophan from 'ophan/ng';
+
+type Module = {
+    name: string,
+    start: number,
+    duration?: number,
+};
+
+type Baseline = {
+    name: string,
+    startTime: number,
+    endTime?: number,
+};
+
+type AdvertMetrics = {
+    id: string,
+    isEmpty: boolean,
+    createTime: number,
+    startLoading: number,
+    stopLoading: number,
+    startRendering: number,
+    stopRendering: number,
+    loadingMethod: number,
+    lazyWaitComplete: number,
+};
+
+const performanceLog = {
+    viewId: 'unknown',
+    tags: ([]: Array<string>),
+    modules: ([]: Array<Module>),
+    adverts: ([]: Array<AdvertMetrics>),
+    baselines: ([]: Array<Baseline>),
+};
+const primaryBaseline = 'primary';
+
+// moduleStart() and moduleEnd() can be used for measuring modules ad-hoc,
+// when they don't align to a baseline.
+const moduleStart = (moduleName: string): void => {
+    const timerStart = getCurrentTime();
+    performanceLog.modules.push({
+        name: moduleName,
+        start: timerStart,
+    });
+};
+
+const moduleEnd = (moduleName: string): void => {
+    const timerEnd = getCurrentTime();
+
+    const moduleIndex = performanceLog.modules.findIndex(
+        module => module.name === moduleName
+    );
+
+    if (moduleIndex !== -1) {
+        const module: Module = performanceLog.modules[moduleIndex];
+        module.duration = timerEnd - module.start;
+    }
+};
+
+// updateAdvertMetric() is called whenever the advert timings need to be updated.
+// It may be called multiple times for the same advert, so that we effectively update
+// the object with additional timings.
+const updateAdvertMetric = (
+    advert: Object,
+    metricName: string,
+    metricValue: string | number
+): void => {
+    performanceLog.adverts = performanceLog.adverts.filter(
+        element => advert.id !== element.id
+    );
+    advert.timings[metricName] = metricValue;
+    performanceLog.adverts.push(
+        Object.freeze(
+            Object.assign(
```

Comment:
Since the usage of `updateAdvertMetric` is to increment a single field of an object in the `performanceLogs.adverts` array, the types should probably reflect that. maybe things like `lazyWaitComplete`, the number fields, are all optional.